### PR TITLE
[WIP] crude working clustered setup

### PR DESCRIPTION
**Pull request by chrisns**
this is using the new `docker swarm` style docker-compose.yml
while it does 'work' and you can trivially scale up the cluster, it's a bit gross particularly around how it gets IP addresses so I'm keen for the thoughts of others.
depends on #10 which I've pushed to my own docker hub account for now.

### CODE REVIEW

**Comment by dergraf** [comment]
Thanks for the contributions! Will have a look at it as soon as possible!

File: bin/vernemq.sh [hunk 1]
```diff
@@ -21,10 +20,10 @@ env | grep DOCKER_VERNEMQ | grep -v DISCOVERY_NODE | cut -c 16- | tr '[:upper:]'
 
 echo "erlang.distribution.port_range.minimum = 9100" >> /etc/vernemq/vernemq.conf
 echo "erlang.distribution.port_range.maximum = 9109" >> /etc/vernemq/vernemq.conf
-echo "listener.tcp.default = ${IP_ADDRESS}:1883" >> /etc/vernemq/vernemq.conf
-echo "listener.ws.default = ${IP_ADDRESS}:8080" >> /etc/vernemq/vernemq.conf
-echo "listener.vmq.clustering = ${IP_ADDRESS}:44053" >> /etc/vernemq/vernemq.conf
-echo "listener.http.metrics = ${IP_ADDRESS}:8888" >> /etc/vernemq/vernemq.conf
+echo "listener.tcp.default = 0.0.0.0:1883" >> /etc/vernemq/vernemq.conf
+echo "listener.ws.default = 0.0.0.0:8080" >> /etc/vernemq/vernemq.conf
+echo "listener.vmq.clustering = 0.0.0.0:44053" >> /etc/vernemq/vernemq.conf
```
Comment (dergraf) [inline bin/vernemq.sh]: `listener.vmq.clustering` should point to a specific IP Address as this one is also used by other cluster nodes to establish the remote message bus.

File: bin/vernemq.sh [hunk 2]
```diff
@@ -63,6 +62,11 @@ trap 'kill ${!}; sigterm_handler' SIGTERM
 /usr/sbin/vernemq start
 pid=$(ps aux | grep '[b]eam.smp' | awk '{print $2}')
 
+if env | grep -q "DOCKER_VERNEMQ_DISCOVERY_NODE"; then
+    DOCKER_VERNEMQ_DISCOVERY_NODE=$(getent hosts ${DOCKER_VERNEMQ_DISCOVERY_NODE} | awk '{ print $1 }' | head -n 1)
+    wait-for-it.sh ${IP_ADDRESS}:44053 ${DOCKER_VERNEMQ_DISCOVERY_NODE}:44053 && vmq-admin cluster join discovery-node=VerneMQ@${DOCKER_VERNEMQ_DISCOVERY_NODE}
```
Comment (dergraf) [inline bin/vernemq.sh]: Sounds like a great idea! :)

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+version: "3"
+services:
+  mqtt_master:
+    image: chrisns/docker-vernemq:addwaitforit
```
Comment (dergraf) [inline docker-compose.yml]: I guess this one should point to `erlio/docker-vernemq` in the end, right? 
Same on line 20

Comment (chrisns) [inline docker-compose.yml]: yeah, totally.
*if* my changes get merged then I'll destroy my `chrisns/docker-vernemq`

File: bin/vernemq.sh (hunk 1 above)
Comment (chrisns) [inline bin/vernemq.sh]: thats if you use `--network host` in my clustered example each node would have its own ip in the docker overlay network

**Comment by chrisns** [comment]
The main thing I really don't like with how this all works is not being able to use hostnames because vernemq complains that they're not valid FQDNs. the thing of trying to figure out the right IP address to listen on is just gross. 😢 
I guess this should maybe iterate through every IP on every interface and setup a listener maybe

**Comment by chrisns** [comment]
following earlier thoughts, I've now refactored so there isn't a need for a separate master in the stack, the swarm containers discover each other.

I'd still rather use the hostnames/container names than the ip addresses since IP addresses may be reused, and I'm not sure how vernemq would handle a cluster node going away and a new one coming back on the same ip?

**Comment by chrisns** [comment]
further.. I've added a travis build to prove the stuff I've done on clustering
https://travis-ci.org/chrisns/docker-vernemq/builds/192748687

**Comment by chrisns** [comment]
any thoughts on this?

**Comment by dergraf** [comment]
Generally ok. Do you know how this works for people not using swarm, e.g. plain docker or kubernetes?

**Comment by chrisns** [comment]
it should work with plain docker, docker-compose (2), kubernetes is another beast and would likely require tweaking based on how service and peer discovery works in that world but probably ok

**Comment by dergraf** [comment]
I've seen that you expect that most users require a DNS based service discovery mechanism. I am not sure if we should do that this way.. as we would possibly end up with a shell script that supports many different service discovery mechanisms. Would it make sense to make this somewhat pluggable in a way that you can define the discovery type in an environment variable along the required configs and let the shell script call a different script exactly for the purpose of discovery? Does this make sense?

**Comment by chrisns** [comment]
the original IP based discovery still works, along with a new `PEER_DISCOVERY_NAME` which yes is round-robin DNS based.
I think you're probably right, though it's really at the limit of what should really be refactored into a not-bash so that it can have tests and other nice things.
If that sounds like something you would merge let me know what language you'd accept it in and I'll try and put it together.

**Comment by dergraf** [comment]
Hi @chrisns 
Sorry for not being more responsive.
If you want to contribute something that could be easily extended with other discovery mechanism or maybe already includes the most important ones I'd be happy to merge your PR.

Personally I'd favor something similar to https://github.com/aweber/rabbitmq-autocluster. However, I don't have a lot of time at the moment. But yeah I don't mind if it is something else.

However if you decide to contribute such a thing it would be great if you could help us out with the maintenance and support. What do you think? Should we have a chat over at Slack at some point and discuss some possibilities?

**Comment by dergraf** [comment]
@chrisns what is the status here? Should we close it in favour of a future more generic solution?

**Comment by chrisns** [comment]
yeah, sure - sorry I've not had time to do anything on this yet with your suggestions

**Comment by rokka-n** [comment]
Discovery based on DNS would work just fine in most case, most container schedules have some flavor of DNS, loadbalancers/proxy or combination.

Hardcoding or using raw IP addresses is definitely gross :)

**Comment by dergraf** [comment]
Gross is a strong word! ;) 
But you're right in the context of containers it isn't appropriate.

**Pull request closed without merging**